File: FEBRUARY 24/09 February/368. Largest Divisible Subset LC.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size();
        vector<pair<int,vector<int>>> dp(n);
        sort(nums.begin(), nums.end());
        for(int i=0; i<n; i++)
            dp[i] = {1,{nums[i]}};
        for(int i=1; i<n; i++)
        {
            for(int j=0; j<i; j++)
            {
                if(nums[i]%nums[j]==0 && dp[i].first<dp[j].first+1)
                {
                    dp[i].first = dp[j].first + 1;
                    dp[i].second = dp[j].second;
                    dp[i].second.push_back(nums[i]);
                }
            }
        }
        int subsetSize = 1, divisibleSubset = 0;
        for(int i=0; i<n; i++)
        {
            if(dp[i].second.size()>subsetSize)
            {
                subsetSize++;
                divisibleSubset = i;
            }
        }
        return dp[divisibleSubset].second;
    }
};

File: APRIL 2025/06 April/368. Largest Divisible Subset LC.java
class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int N = nums.length;
        int[] subsetSize = new int[N+5];
        int[] last = new int[N];

        Arrays.fill(subsetSize, 1);
        
        Arrays.sort(nums);
        int maxSubsetSize = 1, subsetLastIndex=0;
        for(int i=0; i<N; i++)
        {
            last[i] = i;
            for(int j=0; j<i; j++)
            {
                if(nums[i]%nums[j]==0 && subsetSize[i]<subsetSize[j]+1)
                {
                    subsetSize[i] = subsetSize[j]+1;
                    last[i] = j;
                }
            }
            if(subsetSize[i]>maxSubsetSize)
            {
                maxSubsetSize = subsetSize[i];
                subsetLastIndex = i;
            }
        }
        

        List<Integer> subset = new ArrayList<>();
        while(subsetLastIndex!=last[subsetLastIndex])
        {
            subset.add(nums[subsetLastIndex]);
            subsetLastIndex = last[subsetLastIndex];
        }

        subset.add(nums[subsetLastIndex]);
        Collections.reverse(subset);
        
        return subset;
    }
}
